Use IVF_PQ for GPU index build for large datasets

mayya-sharipova opened the pull request:
Use IVF_PQ fallback for insufficient GPU memory

Add adaptive fallback to IVF_PQ algorithm when GPU memory is insufficient for
NN_DESCENT. Include distance type awareness to avoid Cosine distance with
IVF_PQ (unsupported in CUVS 25.10). Add TODO for CUVS 25.12+ upgrade.

Use IVF_PQ algorithm for GPU index building for large dataset (>= 5M vectors).
Temporarily add a factory for calculating IVF_PQ params (to be 
removed with  CUVS 25.12+ upgrade.)

mayya-sharipova committed:
Use IVF_PQ for GPU index build for large datasets

Use IVF_PQ algorithm for GPU index building for large dataset (>= 1M vectors).
Temporarily add a factory for calculating IVF_PQ params.
Also skip estimation of needed memory when IVF_PQ is used.

mayya-sharipova committed:
Update docs/changelog/137126.yaml

mayya-sharipova commented:
With this params (1M byte vectors): 

@achirkin Notice how here when switching from NN_DESCENT, we got worse graph building time, and more dense graphs.

gist: 1_000_000 docs; 960 dims; euclidean metric
index_type |  force_merge_time (ms) | QPS1 seg | recall1 seg
-- |--: | --: | --:
cpu | 130129 | 421 | 0.91
gpu NN_DESCENT | 20643 | 467 | 0.92
gpu IVF_PQ | 36536 | 149 | 1

</div></b>

mayya-sharipova committed:
Use IVF_PQ fallback for insufficient GPU memory

Add adaptive fallback to IVF_PQ algorithm when GPU memory is insufficient for
NN_DESCENT. Include distance type awareness to avoid Cosine distance with
IVF_PQ (unsupported in CUVS 25.10). Add TODO for CUVS 25.12+ upgrade.

ldematte reviewed (changes requested):
The params factory looks good, but I have reservations about the resource manager changes. Let's see if we can figure out a more robust/cleaner way!

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into gpu-ivf-pq

mayya-sharipova committed:
Add memory-aware algorithm selection for GPU HNSW vector indexing

- Implement automatic switching to IVF_PQ algorithm when NN_DESCENT requires
  more GPU memory than available.
- Cache GPU total memory in the GPUSupport module

mayya-sharipova committed:
Merge branch 'main' into gpu-ivf-pq

mayya-sharipova commented on x-pack/plugin/gpu/src/main/java/org/elasticsearch/xpack/gpu/codec/CuVSIvfPqParamsFactory.java:
```diff
@@ -0,0 +1,141 @@
+        double kmeansTrainsetFraction = Math.clamp(1.0 / Math.sqrt(nRows * 1e-5), minKmeansTrainsetFraction, maxKmeansTrainsetFraction);
+
+        // Calculate number of probes based on number of lists
+        int nProbes = Math.round((float) (Math.sqrt(nLists) / 20.0 + 4.0));
```
Addressed in fec345fa5d8f8ea04195d757cc7ccfdef7fd4a65

ChrisHegarty reviewed (approved):
LGTM

ldematte reviewed (commented):
looks good, thanks!
I have a couple of nits and one bigger ask for the GPUSupport class.
That could be done in a follow-up maybe, even though I'd prefer we get it right straight away.

ldematte reviewed (approved):
I am approving this to avoid blocking it, so you are free to decide if you want to address the issue with GPUSupport here or in a follow-up.

mayya-sharipova committed:
Refactor GPUSupport with static initializations

- Simplify GPUSupport.isSupported() by removing unused logError parameter
- Use Holder pattern for thread-safe static initialization of GPU info
- Consolidate duplicate GPU initialization logic into single initializeGpuInfo() method
- Mark two tests @AwaitsFix that fail due to static initialization with mocked GPU provider
- Extract GPU_COMPUTATION_MEMORY_FACTOR constant to CuVSResourceManager interface
- Remove unnecessary debug logging guard in ES92GpuHnswVectorsWriter

mayya-sharipova commented on x-pack/plugin/gpu/src/main/java/org/elasticsearch/xpack/gpu/codec/ES92GpuHnswVectorsWriter.java:
```diff
@@ -350,14 +358,52 @@ private CagraIndexParams createCagraIndexParams(VectorSimilarityFunction similar
+                long requiredMemoryForNnDescent = CuVSResourceManager.estimateNNDescentMemory(numVectors, dims, dataType);
+                if (requiredMemoryForNnDescent > totalDeviceMemory) {
+                    useIvfPQ = true;
+                    if (logger.isDebugEnabled()) {
```
Addressed in ee0907f099686bd13f23e744a91e03e0b536f9af

mayya-sharipova committed:
Merge branch 'main' into gpu-ivf-pq

mayya-sharipova committed:
Add AwaitsFix to GPUPluginInitializationIT tests

ldematte committed:
Fix tests

mayya-sharipova committed:
Merge branch 'main' into gpu-ivf-pq

ldematte committed:
Merge branch 'main' into gpu-ivf-pq

mayya-sharipova committed:
Merge branch 'main' into gpu-ivf-pq
